src/Features/CSharp/Portable/ExtractMethod/CSharpSelectionValidator.cs

```diff
@@ -20,538 +18,574 @@
-                        : selectionInfo.LastTokenInFinalSpan.FullSpan.End;
+            // set final span
+            var start = selectionInfo.FirstTokenInOriginalSpan == selectionInfo.FirstTokenInFinalSpan
+                ? Math.Min(selectionInfo.FirstTokenInOriginalSpan.SpanStart, selectionInfo.OriginalSpan.Start)
+                : selectionInfo.FirstTokenInFinalSpan.FullSpan.Start;
 
-        return selectionInfo.With(s => s.FinalSpan = GetAdjustedSpan(text, TextSpan.FromBounds(start, end)));
-    }
+            var end = selectionInfo.LastTokenInOriginalSpan == selectionInfo.LastTokenInFinalSpan
+                ? Math.Max(selectionInfo.LastTokenInOriginalSpan.Span.End, selectionInfo.OriginalSpan.End)
+                : selectionInfo.LastTokenInFinalSpan.FullSpan.End;
 
-    public override bool ContainsNonReturnExitPointsStatements(IEnumerable<SyntaxNode> jumpsOutOfRegion)
-        => jumpsOutOfRegion.Where(n => n is not ReturnStatementSyntax).Any();
+            return selectionInfo with
+            {
+                FinalSpan = GetAdjustedSpan(text, TextSpan.FromBounds(start, end)),
+            };
+        }
 
-    public override IEnumerable<SyntaxNode> GetOuterReturnStatements(SyntaxNode commonRoot, IEnumerable<SyntaxNode> jumpsOutOfRegion)
-    {
-        var returnStatements = jumpsOutOfRegion.Where(s => s is ReturnStatementSyntax);
+        public override bool ContainsNonReturnExitPointsStatements(ImmutableArray<SyntaxNode> jumpsOutOfRegion)
+            => jumpsOutOfRegion.Any(n => n is not ReturnStatementSyntax);
 
-        var container = commonRoot.GetAncestorsOrThis<SyntaxNode>().Where(a => a.IsReturnableConstruct()).FirstOrDefault();
-        if (container == null)
-            return [];
+        public override IEnumerable<SyntaxNode> GetOuterReturnStatements(SyntaxNode commonRoot, IEnumerable<SyntaxNode> jumpsOutOfRegion)
+        {
+            var returnStatements = jumpsOutOfRegion.Where(s => s is ReturnStatementSyntax);
 
-        var returnableConstructPairs = returnStatements.Select(r => (r, r.GetAncestors<SyntaxNode>().Where(a => a.IsReturnableConstruct()).FirstOrDefault()))
-                                                       .Where(p => p.Item2 != null);
+            var container = commonRoot.GetAncestorsOrThis<SyntaxNode>().Where(a => a.IsReturnableConstruct()).FirstOrDefault();
+            if (container == null)
+                return [];
 
-        // now filter return statements to only include the one under outmost container
-        return returnableConstructPairs.Where(p => p.Item2 == container).Select(p => p.Item1);
-    }
+            var returnableConstructPairs = returnStatements.Select(r => (r, r.GetAncestors<SyntaxNode>().Where(a => a.IsReturnableConstruct()).FirstOrDefault()))
+                                                           .Where(p => p.Item2 != null);
 
-    public override bool IsFinalSpanSemanticallyValidSpan(
-        SyntaxNode root, TextSpan textSpan,
```

Comment:
redundant to pass 'root'.  base class already holds onto it through SemanticDocument.